### DISCUSSION TITLE
Feature class Styles

### DISCUSSION THREAD

**Discussion by sunshine0576** [ngageoint/geopackage-android#55] [Ideas]
Please fill out as much known and relevant information as possible.

#### Version Information:

  * GeoPackage Android Version: 3.2

非常开心能够在新版本中加入样式表，但是我们在实际应用的过程中，通常不会每条要素都有一个样式的，大多数的情况下，都是根据分类来定义几种样式。后续会不会加入这些选项

**Comment by sunshine0576** [comment]
It's great to be able to add stylesheets to the new version, but in practice, we don't always have one style for every element. In most cases, we define several styles according to classification. Will these options be added later?

**Comment by bosborn** [comment]
I'm not sure what you mean by "classification", but you can set defaults for an entire feature table optionally based upon the geometry type.

Here is our [Feature Style extension spec](http://ngageoint.github.io/GeoPackage/docs/extensions/feature-style.html)

And a few examples to set table defaults [here](https://github.com/ngageoint/geopackage-android/blob/33568525f69d62b1c93c41c519e7518b40d1df5e/geopackage-sdk/src/androidTest/java/mil/nga/geopackage/test/GeoPackageExample.java#L2078) and [here](https://github.com/ngageoint/geopackage-android/blob/33568525f69d62b1c93c41c519e7518b40d1df5e/geopackage-sdk/src/androidTest/java/mil/nga/geopackage/test/GeoPackageExample.java#L2143).

There is currently no way to set a style for feature rows based upon a specified column and value.  I'm not sure if that's what you are asking for, but it's not a bad idea for an enhancement.

We'll also be watching the OGC as they develop an official style extension.  But this could take awhile.

**Comment by bosborn** [comment]
Defining a feature table column, an operation (=, <, etc), and value would be useful.  As well as an optional zoom range.

**Comment by osedok** [comment]
Hi Brian, Thank you for providing updated versions of the libraries. This is great news.
Regarding:
"There is currently no way to set a style for feature rows based upon a specified column and value. I'm not sure if that's what you are asking for, but it's not a bad idea for an enhancement."

If this can be considered that would be fantastic addition...

Kind Regards,
Andrzej

**Comment by sunshine0576** [comment]
Implementing expressions can be difficult. A considerable amount of computing is required on mobile devices. But defining rules through fields is a good choice. Can solve most of the applications.

**Comment by sunshine0576** [comment]
That's what I did in my project.

public class SymbolizerFeatureTiles extends DefaultFeatureTiles {
    public SymbolizerFeatureTiles(Context context, FeatureDao featureDao) {
        super(context, featureDao);
    }
    @Override
    protected FeatureStyle getFeatureStyle(FeatureRow featureRow, GeometryType geometryType) {
        int styleID=getFeatureStyleID(featureRow);
        FeatureStyle featureStyle = null;        
        if (featureTableStyles != null) {
            featureStyle = featureTableStyles.getFeatureStyle(styleID, geometryType);
        }
        return featureStyle;
    }
    private String fieldClassName="class";    
    private String[] fielsKeys=new String[]{"default","highway","railway","metro"};
    private int getFeatureStyleID(FeatureRow featureRow) {
        String key=featureRow.getValueString(fieldClassName);
        for (int i = 0; i <fielsKeys.length ; i++) {
            if(fielsKeys[i].equals(key)){
                return i;
            }
        }
        //第一个为默认样式
        return 0;
    }
}

**Comment by bosborn** [comment]
_Converted from an Issue to a Discussion_
